design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+
+## VM Naming
+
+By default, VM names are generated from the VMPool’s name by appending the VMPool’s name with a sequential unique postfix. This is similar to how pods are generated from a StatefulSet’s name.
+
+## State Preservation during Scale-in
+
+During scale-in when the disruptionStrategy is set to `Proactive` with `ScaleInStatePreservation=Offline`, the VM’s being removed from the pool will have their PVC state preserved. In order to ensure on the next scale-out event that VMs using the exact same state are started again, the previous VM names will be reused during scale-out. This is similar in concept to how a StatefulSet uses predictable sequential names.
+
+When disruptionStrategy is set to `Proactive` with `ScaleInStatePreservation=Disabled`, all PVC state will be completely removed from VMs during scale-in and reprovisioned during scale-out.
+
+## Handling of Annotations and Labels
+
+VMs will inherit the Labels/Annotations from the VMPool.
+
+VMIs will inherit the Labels/Annotations from the VMITemplate section of a VMConfig in the same fashion that VMIs inherit Labels/Annotations from the VMITemplate of a VM. 
+
+## Handling Persistent Storage
+
+Usage of a DataVolumeTemplate within a VMConfig will result in unique persistent storage getting created for each VM within a VMPool. The DataVolumeTemplate name will have the VM’s sequential postfix appended to when the VM is created from the VMConfig which makes each VM a completely unique stateful workload.
+
+## Handling Unique VM CloudInit and ConfigMap Volumes at Scale
+
+By default, any secrets or configMaps references in a VMConfig’s Volume section will be used directly as is, without any modification to the naming. This means if you specify a secret in a CloudInitNoCloud volume within a VMConfig, that every VM instance spawned from that VMConfig will get the exact same secret used for their cloud-init user data.
+
+This default behavior can be modified by setting the **AppendPostfixToSecretReferences** and **AppendPostfixToConfigMapReferences** booleans to true on the VMPool spec. When these booleans are enabled, references to secret and configMap names will have the VM’s sequential postfix appended to the secret and configmap name. This allows someone to pre-generate unique per VM secret and configMap data for a VMPool ahead of time in a way that will be predictably assigned to VMs within the VMPool.
+
+## Autohealing
+
+When managing VMs at large scale, it is useful to automate the recovery of VMs which continually fail to reach a ready state. This automatically fixes scenarios where a VM’s state has somehow been corrupted and needs to be completely refreshed.
+
+By setting `AutohealingStrategy=ReprovisionOnFailure` on the VMPool’s spec, VMs which reach an unexpected failure state will automatically be completely deleted (including persistent storage) and re-provisioned. This allows for auto recovery in application scenarios that can withstand such an action.
```

Comment:
We should probably also have some sort of CrashLoopBackoff just to make sure we don't contribute to any strain that might cause provisioning to fail in the first place.

Reply:
yup, i'll note that.